addons/hr_payroll_account/models/hr_payroll_account.py

```diff
@@ -0,0 +1,155 @@
+#-*- coding:utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+from odoo.tools import float_compare, float_is_zero
+
+
+class HrPayslipLine(models.Model):
+    _inherit = 'hr.payslip.line'
+
+    def _get_partner_id(self, credit_account):
+        """
+        Get partner_id of slip line to use in account_move_line
+        """
+        # use partner of salary rule or fallback on employee's address
+        partner_id = self.salary_rule_id.register_id.partner_id.id or self.slip_id.employee_id.address_home_id.id
+        if credit_account and self.salary_rule_id.register_id.partner_id or self.salary_rule_id.account_credit.internal_type in ('receivable', 'payable'):
+            return partner_id
+        elif self.salary_rule_id.register_id.partner_id or self.salary_rule_id.account_debit.internal_type in ('receivable', 'payable'):
+            return partner_id
+        return False
+
+
+class HrPayslip(models.Model):
+    _inherit = 'hr.payslip'
+
+    date_account = fields.Date(string='Date Account', states={'draft': [('readonly', False)]}, readonly=True, help="Keep empty to use the period of the validation(Payslip) date.", oldname="date")
+    journal_id = fields.Many2one('account.journal', string='Salary Journal', states={'draft': [('readonly', False)]}, readonly=True, required=True, default=lambda self: self.env['account.journal'].search([('type', '=', 'general')], limit=1).id)
+    move_id = fields.Many2one('account.move', string='Accounting Entry', readonly=True, copy=False)
+
+    @api.model
+    def create(self, vals):
+        if 'journal_id' in self.env.context:
+            vals.update({'journal_id': self.env.context.get('journal_id')})
+        return super(HrPayslip, self).create(vals)
+
+    @api.multi
+    def onchange_contract_id(self, date_from, date_to, employee_id=False, contract_id=False):
+        res = super(HrPayslip, self).onchange_contract_id(date_from=date_from, date_to=date_to, employee_id=employee_id, contract_id=contract_id)
+        journal_id = self.env['hr.contract'].browse(contract_id).journal_id.id
+        res['value'].update({'journal_id': journal_id})
+        return res
+
+    @api.multi
+    def cancel_sheet(self):
+        AccountMove = self.env['account.move']
+        move_ids = []
+        move_to_cancel = []
+        for slip in self.filtered('move_id'):
+            move_ids.append(slip.move_id.id)
+            if slip.move_id.state == 'posted':
+                move_to_cancel.append(slip.move_id.id)
+        AccountMove.browse(move_to_cancel).button_cancel()
+        AccountMove.browse(move_ids).unlink()
+        return super(HrPayslip, self).cancel_sheet()
```

Comment:
this cancel_sheet can improved like this way: 
	 moves = self.mapped('move_id')
         moves_cancel = moves.filtered(lambda m: m.state == 'posted').button_cancel()
         moves.unlink()
         return super(HrPayslip, self).cancel_sheet()